---
title: "Integrated Doublet Cell Detection in Single-Cell Multiome (scRNA-seq & scATAC-seq)"
author: Herv√© Ishimwe
output:
  pdf_document: default
  html_document:
    theme: spacelab
    df_print: kable
    toc: true
    toc_float: 
      collapsed: true
---

```{r setup, include=FALSE}
all_times <- list()  # store the time for each chunk
knitr::knit_hooks$set(time_it = local({
  now <- NULL
  function(before, options) {
    if (before) {
      now <<- Sys.time()
    } else {
      res <- difftime(Sys.time(), now)
      all_times[[options$label]] <<- res
    }
  }
}))
knitr::opts_chunk$set(
  message = FALSE,
  warning = FALSE,
  error = FALSE,
  cache.lazy = FALSE,
  time_it = TRUE
)
```

This vignette introduces the voting approach for integrated doublet cell detection in single-cell multiome. It is divided into three main sections:

1. `Seurat` workflow which consists of:
  - RNA and ATAC preprocessing and dimension reductions
  - Multimodal clustering following: [\textcolor{blue}{Weighted Nearest Neighbor Analysis of 10x multiome} data](https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis#wnn-analysis-of-10x-multiome-rna-atac)
  - Cell type annotation of the dataset following: [\textcolor{blue}{Multimodal reference mapping Vignette}](https://satijalab.org/seurat/articles/multimodal_reference_mapping.html)

2. Running the doublet cell detection tools for each profile:
  - `scATAC-seq` profile: we run the `ArchR` following the [\textcolor{blue}{ArchR tutorial}](https://www.archrproject.com/articles/Articles/tutorial.html)
  - `scRNA-seq` profile: we run two doublet cell detection tools, 
    + `DoubletFinder` following [\textcolor{blue}{DoubletFinder's example code for 'real-world' applications}](https://github.com/chris-mcginnis-ucsf/DoubletFinder?tab=readme-ov-file)
    + `scDblFinder` following [\textcolor{blue}{scDblFinder tutorial}](https://bioconductor.org/packages/release/bioc/vignettes/scDblFinder/inst/doc/scDblFinder.html)

3. Classification of cell clusters and cells. This involves aggregating the information from doublet cell detection tools and cell type annotation then, classify cell clusters and cells as doublets or not. 

Required libraries to run this vignette:

```{r packages, cache=FALSE} 
library(Seurat)
library(SeuratDisk)
library(Signac)
library(EnsDb.Hsapiens.v86)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(DoubletFinder)
library(patchwork)
library(dplyr)
library(scDblFinder)
library(ArchR)
ArchR::installExtraPackages()
```

In this vignette, we use a sample from a PBMC single-cell multiome dataset  with `scRNA-seq` and `scATAC-seq` profiles. Datasets similar to `10x Genomics's Multiome` can follow a similar format.

\newpage

## Seurat Workflow

```{r wd, echo=FALSE}
working_directory <- paste("~/Desktop/Thesis/OP_Multiome/samples/DU19-01S0004389_7")
```

```{r seurat}
setwd(working_directory)
sample_matrix <- Read10X_h5("filtered_feature_bc_matrix.h5") 
frag_file <- "atac_fragments.tsv.gz"

# creating the seurat object

# extracting RNA counts
seurat_object <- CreateSeuratObject(counts = sample_matrix$`Gene Expression`)
seurat_object[["percent.mt"]] <- PercentageFeatureSet(seurat_object, pattern = "^MT-")

# extracting ATAC counts
atac_counts <- sample_matrix$Peaks
grange.counts <- StringToGRanges(rownames(atac_counts), sep = c(":", "-"))
grange.use <- seqnames(grange.counts) %in% standardChromosomes(grange.counts)
atac_counts <- atac_counts[as.vector(grange.use), ]
annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)
seqlevelsStyle(annotations) <- 'UCSC'
genome(annotations) <- "hg38"
chrom_assay <- CreateChromatinAssay(
  counts = atac_counts,
  sep = c(":", "-"),
  genome = 'hg38',
  fragments = frag_file,
  min.cells = 10,
  annotation = annotations
)
seurat_object[["ATAC"]] <- chrom_assay

seurat_object <- NormalizeData(seurat_object, verbose = FALSE)
seurat_object <- ScaleData(seurat_object, features = rownames(seurat_object), 
                           verbose = FALSE)
seurat_object
```

### Preprocessing & dimension reduction

Here we perform assay specific dimension reduction: `PCA` for `scRNA-seq` data and `LSI` for `scATAC-seq` data.

```{r dimensions}
dimensions <- list("dims_pca" = 1:30, "dims_lsi" = 2:30)
```

```{r pca}
DefaultAssay(seurat_object) <- "RNA"
seurat_object <- SCTransform(seurat_object)
seurat_object <- RunPCA(seurat_object)
seurat_object <- RunUMAP(seurat_object, dims = dimensions$dims_pca, 
                         reduction.name = 'umap.rna', reduction.key = 'rnaUMAP_', 
                         verbose = FALSE)
```

```{r lsi}
DefaultAssay(seurat_object) <- "ATAC"
seurat_object <- RunTFIDF(seurat_object)
seurat_object <- FindTopFeatures(seurat_object, min.cutoff = 'q0')
seurat_object <- RunSVD(seurat_object)
seurat_object <- RunUMAP(seurat_object, reduction = 'lsi', dims = dimensions$dims_lsi, 
                         reduction.name = "umap.atac", reduction.key = "atacUMAP_")
```

###  Multimodal clustering

We rely on `Seurat`'s multi-modal clustering which incorporates the information from both profiles to form a weighted nearest neighbor graph then we find the cluster in this graph.

```{r clusters}
seurat_object <- FindMultiModalNeighbors(seurat_object, 
                                         reduction.list = list("pca", "lsi"), 
                                         dims.list = dimensions, verbose = FALSE)
seurat_object <- RunUMAP(seurat_object, nn.name = "weighted.nn", 
                         reduction.name = "wnn.umap", 
                         reduction.key = "wnnUMAP_", verbose = FALSE)
seurat_object <- FindClusters(seurat_object, graph.name = "wsnn", algorithm = 3, 
                              verbose = FALSE)
```

We visualize the clusters on the [\textcolor{blue}{UMAP}](https://arxiv.org/pdf/1802.03426.pdf?source=post_page) one for each profile (RNA and ATAC) and another for the weighted nearest neighbor which is a multimodal representation.

```{r clusterplots, fig.width=10, fig.height=10, echo = FALSE}
cluster_plot_wnn <- DimPlot(seurat_object, reduction = "wnn.umap", 
                            group.by = "seurat_clusters", label = TRUE, 
                            label.size = 3, repel = TRUE) + ggtitle("WNN Clusters: Sample_4389")
cluster_plot_wnn
```

### Cell Type Annotation 

We use `Seurat` mapping and annotating functions to predict the cell types using a reference dataset from the following paper: [\textcolor{blue}{(Hao, Hao et al, Cell 2021)}](https://www.sciencedirect.com/science/article/pii/S0092867421005833?via%3Dihub) 

```{r loadReference}
reference <- LoadH5Seurat("~/Desktop/Thesis/data/pbmc_multimodal.h5seurat")
```

Since we have gene expression information, we can follow this [\textcolor{blue}{Multimodal reference mapping Vignette}](https://satijalab.org/seurat/articles/multimodal_reference_mapping.html) to identify anchors then transfer cell types to our dataset.

```{r annotation}
DefaultAssay(seurat_object) <- "RNA"
anchors <- FindTransferAnchors(
  reference = reference,
  query = seurat_object,
  normalization.method = "SCT",
  reference.reduction = "spca",
  dims = dimensions$dims_pca
)

seurat_object <- MapQuery(
  anchorset = anchors,
  query = seurat_object,
  reference = reference,
  refdata = list(
    celltype.l1 = "celltype.l1", ## levels of accuracy
    celltype.l2 = "celltype.l2",
    celltype.l3 = "celltype.l3",
    predicted_ADT = "ADT"
  ),
  reference.reduction = "spca",
  reduction.model = "wnn.umap"
)
```

#### Selecting Cell Type Annotation Level 

For each annotation level of granularity, we have prediction scores which are on a scale from `0` to `1` where a higher score corresponds to the level of confidence of the cell type annotation. We want to select a robust annotation level which has the highest mean (skewed towards 1) and a small standard deviation.

```{r annotLevel, echo=FALSE}
annotation_stats <- data.frame(
  "granularity level" = c("1", "2", "3"),
  "mean" = c(mean(seurat_object@meta.data$predicted.celltype.l1.score), 
             mean(seurat_object@meta.data$predicted.celltype.l2.score), 
             mean(seurat_object@meta.data$predicted.celltype.l3.score)),
  "standard_deviation" = c(sd(seurat_object@meta.data$predicted.celltype.l1.score), 
                           sd(seurat_object@meta.data$predicted.celltype.l2.score), 
                           sd(seurat_object@meta.data$predicted.celltype.l3.score)) 
)
annotation_stats
```

```{r selectingAnnot, echo=FALSE}
max_mean <- max(annotation_stats$mean)
min_sd <- min(annotation_stats$standard_deviation)

annotation_stats <- annotation_stats %>% filter(annotation_stats$mean == max_mean & annotation_stats$standard_deviation == min_sd)
sprintf("Selected level of granularity is : % s. ", annotation_stats$granularity.level)
```

```{r annotationplots, fig.width=10, fig.height=10, echo = FALSE}
l1_plot <- DimPlot(seurat_object, reduction = "wnn.umap", 
                   group.by = "predicted.celltype.l1", label = TRUE, 
                   label.size = 5, repel = TRUE) + ggtitle("Level 1 Cell Type Annotation: Sample_4389")
l1_plot
```

## Doublet Cell Detection Tools 

### ArchR 

Here we follow the [\textcolor{blue}{brief ArchR tutorial}](https://www.archrproject.com/articles/Articles/tutorial.html) to create `ArrowFiles` from the fragments file then generate doublet scores on the cells identified followed by the creation of the `ArchRProject`.

```{r archRscores, results='hide'}
setwd(working_directory)
set.seed(1)
addArchRThreads(threads = 16) 
addArchRGenome("hg38")

ArrowFiles <- createArrowFiles(
  inputFiles = frag_file,
  sampleNames = "pbmc",
  addTileMat = TRUE,
  addGeneScoreMat = TRUE
)

arch_results <- addDoubletScores(input = ArrowFiles)
sample_archRProject <- ArchRProject(ArrowFiles = ArrowFiles)
```

Since `ArchR` doesn't have generate labels to indicate which cells classified as doublets and which are not, we rely on the functionality of the `ArchR` function [\textcolor{blue}{filterDoublets}](https://www.archrproject.com/reference/filterDoublets.html) which has an argument, `filterRatio`, that determines the ratio of doublets to filter out. Then using the default value `filterRatio = 1`, we estimated the number of doublets identified using the following formula $$\frac{filterRatio * (number \space of \space cells)^2}{100000}$$ From this, we label the cells with the lowest [\textcolor{blue}{doublet enrichment}](https://www.archrproject.com/bookdown/inferring-scatac-seq-doublets-with-archr.html) generated from the function `addDoubletScores` as doublets (which corresponds to the number of doublets computed earlier) and the rest as singlets.

```{r archRLabels}
arch_enrich <- arch_results[[1]]@listData[["doubletEnrich"]]

## formatting cell ids 
extracted_cell_ids <- names(arch_enrich)
arch_cell_ids <- c()
for (id in extracted_cell_ids) {
  arch_cell_ids <- c(arch_cell_ids, unlist(strsplit(id, split='#', fixed=TRUE))[2])
}

arch_info <- data.frame(
  cell_id = arch_cell_ids,
  arch_enrich = unname(arch_enrich)
)

## labeling the cells using the filtering mechanism of archR (filter ratio = 1)
arch_cell_count <- length(arch_enrich)
arch_doublet_count <- round(arch_cell_count ^ 2 / 100000) 
arch_singlet_count <- arch_cell_count - arch_doublet_count

# sort the cells in increasing order then remove the highest ones (doublets)
arch_info <- arrange(arch_info, arch_enrich)
arch_labels <- vector(mode="character", length = arch_cell_count)
arch_labels <- factor(arch_labels, levels = c('singlet', 'doublet'))
arch_info$arch_label <- arch_labels

## setting the labels 
arch_info$arch_label[1:arch_singlet_count] <- 'singlet'
arch_info$arch_label[(arch_singlet_count + 1): arch_cell_count] <- 'doublet'
arch_info <- subset(arch_info, select=-arch_enrich)
```

In order to generate a visualization similar to those in `Seurat`, we create a new `Seurat` object from the `GeneScoreMatrix` as follow:

```{r archRSeurat}
geneScoreMatrix <- ArchR::getMatrixFromProject(sample_archRProject, 
                                               useMatrix='GeneScoreMatrix')
rownames(geneScoreMatrix) <-getFeatures(sample_archRProject, 
                                        useMatrix = "GeneScoreMatrix", 
                                        select = NULL,ignoreCase = TRUE)

arch_sce <- as(geneScoreMatrix, "SingleCellExperiment")
arch_sce@assays@data@listData$logcounts<-arch_sce@assays@data@listData$GeneScoreMatrix
arch_seurat <- as.Seurat(arch_sce, counts="GeneScoreMatrix")

## LSI dimension reduction (ATAC)
arch_seurat <- RunTFIDF(arch_seurat)
arch_seurat <- FindTopFeatures(arch_seurat, min.cutoff = 'q0')
arch_seurat <- RunSVD(arch_seurat)
arch_seurat <- RunUMAP(arch_seurat, reduction = 'lsi', dims = dimensions$dims_lsi, 
                        reduction.name = "umap.atac", reduction.key = "atacUMAP_", )
```

```{r archRLabelMatching}
## making sure we match the label to the right cell
extracted_names <- names(arch_seurat@active.ident)
geneScore_names <- c()
for (name in extracted_names) {
  geneScore_names <- c(geneScore_names, unlist(strsplit(name, split='#', 
                                                        fixed=TRUE))[2])
}

geneScore_details = data.frame(cell_id = geneScore_names)
geneScore_details <- merge(geneScore_details, arch_info, by ="cell_id", sort = FALSE)
arch_seurat$archR.labels <- geneScore_details$arch_label
```

```{r archRPlots, fig.width=10, fig.height=10, echo = FALSE}
arch_doublets_plot <- DimPlot(arch_seurat, reduction = "umap.atac", 
                               group.by = "archR.labels", label = FALSE, repel = TRUE) 
arch_doublets_plot <- arch_doublets_plot + ggtitle("ArchR Labels: Sample_4389")
arch_doublets_plot
```

### scDblFinder

Here we generate doublet scores and labels for the cells processed in the `seurat` object. We use the random mode for the generation of artificial doublets.

```{r scdblFinder}
sample_sce <- as.SingleCellExperiment(seurat_object)
scdbl_random <- scDblFinder(sample_sce)

seurat_object$scdbl.scores <- scdbl_random$scDblFinder.score
seurat_object$scdbl.labels <- scdbl_random$scDblFinder.class
```

```{r scdblPlots, fig.width=10, fig.height=10, echo = FALSE}
scdbl_doublets_plot <- DimPlot(seurat_object, reduction = "wnn.umap", 
                            group.by = "scdbl.labels", label = FALSE, repel = TRUE) 
scdbl_doublets_plot <- scdbl_doublets_plot + ggtitle("scDblFinder Labels: Sample_4389")
scdbl_doublets_plot 
```

### DoubletFinder 

```{r paramSweep, results='hide'}
sweep.res <- paramSweep_v3(seurat_object, PCs = 1:30, sct = TRUE)
sweep.stats <- summarizeSweep(sweep.res, GT = FALSE)
bcmv <- find.pK(sweep.stats)
pK.set <- unique(sweep.stats$pK)[2]
```

For the estimated number of doublets, we use the estimated doublet cell percentages for `10x Multiome` dataset from this [\textcolor{blue}{source}](https://uofuhealth.utah.edu/huntsman/shared-resources/gcb/htg/single-cell/genomics-10x)

```{r doubletFinder}
doublet_percent_estimate <- 0.09
nExp_poi <- round(doublet_percent_estimate*nrow(seurat_object@meta.data)) 
seurat_object <- doubletFinder_v3(seurat_object, PCs = 1:30, 
                                  pK = as.numeric(as.character(pK.set)), 
                                  nExp = nExp_poi, sct = TRUE)
```

```{r doubletFinderplots, fig.width=10, fig.height=10, echo = FALSE}
df_doublets_plot <- DimPlot(seurat_object, reduction = "wnn.umap", 
                            group.by = "DF.classifications_0.25_0.005_1125",
                            label = FALSE, repel = TRUE) 
df_doublets_plot <- df_doublets_plot + ggtitle("DoubletFinder Labels: Sample_4389")
df_doublets_plot 
```

## Classification of Cells and Cell Clusters

### Aggregating scores 

Then we start extracting the cluster ids, prediction scores, and doublet scores and labels from the `seurat` object, `DoubletFinder` and `scDblFinder` results first because they have the same number of processed cells.

```{r integration}
seurat_info <- data.frame(
  cell_id = c(names(seurat_object@active.ident)),
  cluster_id = c(seurat_object@meta.data$seurat_clusters),
  l1_annotation_score = c(seurat_object@meta.data$predicted.celltype.l1.score),
  doubletFinder_label = c(seurat_object@meta.data$DF.classifications_0.25_0.005_1125),
  scDbl_label = c(seurat_object@meta.data$scdbl.labels)
)
```

```{r minor, echo=FALSE}
seurat_info$doubletFinder_label <- tolower(seurat_info$doubletFinder_label)
```

Then we merged this information from seurat with that from archR. We anticipate cell filtering from both directions. 

```{r merging}
combined_info <- merge(seurat_info, arch_info, by = "cell_id")
```

### Collecting Votes

Now that we have all the scores together, we can start gathering the votes.

#### Cell Quality

For each cell, we compute a `cell-quality-based` vote corresponding to the number of doublet cell detection tools (`ArchR`, `scDblFinder`, `DoubletFinder`) that classified it as a doublet. For each cluster, we count the number of `imbalanced` cells which refers to cells classified as doublet by at least one doublet detection tool.

```{r toolVoting}
cell_count <- length(combined_info$cell_id)
combined_info$cell_quality_vote <- vector(mode="integer", length=cell_count)

cluster_ids <- levels(seurat_object@meta.data[["seurat_clusters"]])
cluster_info <- data.frame(
  "cluster_id" = cluster_ids,
  "cluster_count" = vector(mode="integer", length = length(cluster_ids)),
  "imbalanced_count" = vector(mode="numeric", length = length(cluster_ids))
)

for (id in cluster_ids) {
  cluster_cells <- combined_info %>% filter(combined_info$cluster_id == id)
  cluster_count <- length(cluster_cells$cell_id)
  cluster_info[cluster_info$cluster_id == id, "cluster_count"] <- cluster_count

  imbalanced_count <- 0
  for (cell in cluster_cells$cell_id) {
    cell_info <- cluster_cells[cluster_cells$cell_id == cell,]
    cell_quality_vote <- 0
    
    if (cell_info$doubletFinder_label == 'doublet') {
      cell_quality_vote <- cell_quality_vote + 1
    }
    
    if (cell_info$scDbl_label == 'doublet') {
      cell_quality_vote <- cell_quality_vote + 1
    }
    
    if (cell_info$arch_label == 'doublet') {
      cell_quality_vote <- cell_quality_vote + 1
    }
    
    combined_info[combined_info$cell_id == cell, "cell_quality_vote"] <- cell_quality_vote
    
    if(cell_quality_vote >= 1) {
      imbalanced_count <- imbalanced_count + 1
    }
  }
  
  cluster_info[cluster_info$cluster_id == id, "imbalanced_count"] <- imbalanced_count
}
```

#### Cell Type Annotation

Now that we know which the number of cells that are candidates for doublet classification i.e `imbalanced cells`, we can look at the cell type annotation using the prediction score. Our assumption is that the doublet cells especially heterotypic doublet cells will be annotated with the lowest prediction scores. Therefore, for each cluster, we select the cells with the lowest prediction scores for the cell type annotation vote. The number of cells selected corresponds to the number of imbalanced cells in a given cell cluster computed in the previous step. 

```{r annotationVote}
combined_info$cell_type_vote <- vector(mode="integer", length=cell_count)

for (id in cluster_ids) {
  cluster_cells <- combined_info %>% filter(combined_info$cluster_id == id)
  
  ## arrange the cells by l1 prediction scores in ascending order
  cluster_cells <- arrange(cluster_cells, l1_annotation_score)
  
  imbalanced_count <- cluster_info[cluster_info$cluster_id == id, "imbalanced_count"]
  cluster_count <- cluster_info[cluster_info$cluster_id == id, "cluster_count"]
  
  for (index in 1:cluster_count) {
    cell_id <- cluster_cells$cell_id[index]
    cell_quality_vote <- combined_info[combined_info$cell_id == cell_id, "cell_quality_vote"] 
    
    ## cells with the lowest prediction scores
    if (index < imbalanced_count + 1) {
      combined_info[combined_info$cell_id == cell_id, "cell_type_vote"] <- 1
    } else {
      combined_info[combined_info$cell_id == cell_id, "cell_type_vote"] <- 0
    }
  }
}
```

Now we combine the `cell-quality-based` vote and the `cell-type-based` vote into one vote, `total vote`.

```{r totalVote}
combined_info$total_vote <- combined_info$cell_quality_vote + combined_info$cell_type_vote
```

#### Cluster Classification

Our assumption here is that if a cell has at least two votes, that means that it is likely to be a doublet cell. We refer to such cells as `doublet candidates`

To classify cell clusters, we first compute the percentage of doublet candidates in the dataset as well as in each cluster. Then, we compute a fold change of each cluster's percentage relative to that of the dataset. If the fold change is greater or equal to 3 then all cells in that given cluster is classified as doublets. 

```{r datasetCandidates}
doublet_candidates <- combined_info %>% filter(combined_info$total_vote >= 2)
sample_candidate_percentage <- length(doublet_candidates$cell_id) * 100 / cell_count
sample_candidate_percentage
```

```{r foldChange}
cluster_info$dbt_candidate_percent<- vector(mode="integer", length = length(cluster_ids))
cluster_info$fold_change <- vector(mode="integer", length = length(cluster_ids))

cluster_labels <- vector(mode="character", length= length(cluster_ids))
cluster_labels <- factor(cluster_labels, levels=c('singlet', 'doublet'))
cluster_info$cluster_label <- cluster_labels

for (id in cluster_ids) {
  cluster_cells <- combined_info %>% filter(combined_info$cluster_id == id)
  
  doublet_candidate_count <- 0 
  for (cell in cluster_cells$cell_id) {
    if (combined_info[combined_info$cell_id == cell, "total_vote"] >= 2) {
      doublet_candidate_count <- doublet_candidate_count + 1
    } 
  }
  
  cluster_count <- cluster_info[cluster_info$cluster_id == id, "cluster_count"] 
  dbt_candidate_percent <- doublet_candidate_count * 100 / cluster_count
  cluster_info[cluster_info$cluster_id == id, "dbt_candidate_percent"] <- dbt_candidate_percent
  
  fold_change <- dbt_candidate_percent / sample_candidate_percentage
  cluster_info[cluster_info$cluster_id == id, "fold_change"] <- fold_change
  
  if (fold_change >= 2.5) {
    cluster_info[cluster_info$cluster_id == id, "cluster_label"] <- 'doublet'
  } else {
    cluster_info[cluster_info$cluster_id == id, "cluster_label"] <- 'singlet'
  }
}
```

```{r clusterInfo, echo=FALSE}
cluster_info <- subset(cluster_info, select=-imbalanced_count)
cluster_info
```

Now, we can make the final classification for each cell as follows: if the cell is located in a cluster classified as a doublet cluster or it has a total vote greater than or equal to 3, we classify it as a doublet. If not, it is classified as a singlet.

```{r finalLabel}
vote_alg_label <- vector(mode="character", length= cell_count)
vote_alg_label <- factor(vote_alg_label, levels=c('singlet', 'doublet'))
combined_info$vote_alg_label <- vote_alg_label
for (id in cluster_ids) {
  cluster_label <- cluster_info[cluster_info$cluster_id == id, "cluster_label"]
  cluster_cells <- combined_info %>% filter(combined_info$cluster_id == id)
  
  for (cell in cluster_cells$cell_id) {
    cell_info <- cluster_cells[cluster_cells$cell_id == cell,]
    
    if (cluster_label == 'doublet' || cell_info$total_vote >= 3) {
      combined_info[combined_info$cell_id == cell, "vote_alg_label"] <- 'doublet'
    } else {
      combined_info[combined_info$cell_id == cell, "vote_alg_label"] <- 'singlet'
    }
  }
}
```

```{r finalDoubletPercentage}
voting_doublets <- combined_info %>% filter(combined_info$vote_alg_label == 'doublet')
vote_alg_percentage <- length(voting_doublets$cell_id) * 100 / cell_count
vote_alg_percentage
```

`Note:` the fold change and total vote thresholds can differ for different samples. So try different values and observe which one generates meaningful doublet rates. 

We can visualize these classifications based on the voting algorithm. Since there was filtering, some of the recovered cells in the `Seurat` object will not have a final label, so we label them as `unclassified`.

```{r votesPlot, echo=FALSE}
## recovering the order of cell ids in Seurat objects
seurat_labels <- data.frame(cell_id = c(names(seurat_object@active.ident)))
num_cells <- length(seurat_labels$cell_id)

vote_label <- vector(mode="character", length= num_cells)
vote_label <- factor(vote_label, levels=c('singlet', 'doublet', 'unclassified'))
seurat_labels$vote_label <- vote_label

seurat_labels <- merge(seurat_labels, combined_info, by = "cell_id", 
                       all.x = TRUE, sort = FALSE)

for (cell in seurat_labels$cell_id) {
  cell_info <- seurat_labels[seurat_labels$cell_id == cell, ]
  
  if (is.na(cell_info$vote_alg_label)) {
    seurat_labels[seurat_labels$cell_id == cell, "vote_label"] <- 'unclassified'
  } else {
    if(cell_info$vote_alg_label == 'doublet') {
      seurat_labels[seurat_labels$cell_id == cell, "vote_label"] <- 'doublet'
    } else if (cell_info$vote_alg_label == 'singlet') {
      seurat_labels[seurat_labels$cell_id == cell, "vote_label"] <- 'singlet'
    }
  }
}

seurat_object$vote.label <- seurat_labels$vote_label

final_doublets_plot <- DimPlot(seurat_object, reduction = "wnn.umap", 
                               group.by = "vote.label", label = FALSE, repel = TRUE) 
final_doublets_plot <- final_doublets_plot + ggtitle("Voting Algorithm Labels: Sample_4389")
final_doublets_plot
```

```{r classificationStats, echo=FALSE}
singlet <- seurat_labels %>% filter(seurat_labels$vote_label == 'singlet')
doublet <- seurat_labels %>% filter(seurat_labels$vote_label == 'doublet')
unclassified <- seurat_labels %>% filter(seurat_labels$vote_label == 'unclassified')

percentages <- data.frame(
  label = c('singlet', 'doublet', 'unclassified'),
  percentage = c(length(singlet$cell_id), length(doublet$cell_id),
                 length(unclassified$cell_id))
)
percentages$percentage <- percentages$percentage * 100 / num_cells
percentages
```

\newpage
<details>
  <summary>**Session Info**</summary>
```{r}
sessionInfo()
```
</details>
